### PR TITLE
Consistent random number handling and improve sample_sphere

### PR DESCRIPTION
**Pull request by mapazarr**
`~gammapy.utils.random.sample_sphere` should use astropy `Angle` objects for better handling / less error prone.
TODO:
 - [x] decide if we use `Angle` or `Longitude` and `Latitude` objects
 - [x] update/revise other functions using `sample_sphere`
 - [x] finish applying issue #238 (`random_state` parameters)

### CODE REVIEW

**Comment by mapazarr** [comment]
Implemented. Merge?

**Comment by cdeil** [comment]
I'd prefer it if we use the more explicit `Longitude` and `Latitude` objects from `astropy.coordinates`.
They have convenience methods for angle wrapping.
I've never used them much, so if that doesn't work well for some reason, please at least use `Angle` objects.

File: gammapy/utils/random.py [hunk 1]
```diff
@@ -39,38 +40,32 @@ def sample_sphere(size, lon_range=None, lat_range=None, unit='rad'):
...
     size : int
         Number of samples to generate
-    lon_range : tuple
-        Longitude range (min, max) tuple in range (0, 360)
-    lat_range : tuple
-        Latitude range (min, max) tuple in range (-90, 90)
-    units : {'rad', 'deg'}
-        Units of input range and returned angles
+    lon_range : `~astropy.units.Quantity`, optional
+        Longitude range (min, max) in range (0, 360) deg
```
Comment (cdeil) [inline gammapy/utils/random.py]: Would it be easy to make the code just work for longitudes in the range `(0, 360)` and `(-180, 180)`?
I.e. not have this requirement of `lon_range` being in the range `(0, 360)`?
If yes, please do.

File: gammapy/utils/random.py [hunk 2]
```diff
@@ -39,38 +40,32 @@ def sample_sphere(size, lon_range=None, lat_range=None, unit='rad'):
...
-        raise ValueError('Invalid unit: {0}'.format(unit))
-
     # Convert inputs to internal format (all radians)
     size = int(size)
 
-    if (lon_range is not None) and (unit == 'deg'):
-        lon_range = np.radians(lon_range)
+    if lon_range is not None:
+        lon_unit = lon_range.unit
+        lon_range.to('radian')
```
Comment (cdeil) [inline gammapy/utils/random.py]: @mapazarr – Here I'm not sure what the best API is.
Is it better to always return angles in the same units as the inputs?
What's more intuitive / convenient for callers?

File: gammapy/utils/random.py [hunk 3]
```diff
@@ -39,38 +40,32 @@ def sample_sphere(size, lon_range=None, lat_range=None, unit='rad'):
...
     else:
-        lon_range = 0, 2 * np.pi
+        lon_unit = 'radian'
+        lon_range = Quantity([0., 2*np.pi], lon_unit)
 
-    if (lat_range is not None) and (unit == 'deg'):
-        lat_range = np.radians(lat_range)
+    if lat_range is not None:
+        lat_unit = lon_range.unit
+        lat_range.to('radian')
```
Comment (cdeil) [inline gammapy/utils/random.py]: I think this statement has no effect:
```
In [3]: a = Angle(10, 'degree')

In [4]: a.to('radian')
Out[4]: <Angle 0.17453292519943295 rad>

In [5]: a
Out[5]: <Angle 10.0 deg>
```

Also: the numpy trig functions work with Angle, so no need for unit conversion to radian on input?

**Comment by cdeil** [comment]
Can you please look at the tests for this function and see if they cover the common and some edge cases well, and if not, please add a test or two?

**Comment by mapazarr** [comment]
Sorry, I missed the `Longitude` `Latitude` part of the comment. I tried to use them, but it's not straightforward, and would need deeper debugging, for which I have no time right now. Indeed, `Angle` would be much more appropriate as `Quantity`. I am using `Angle` right now.

File: gammapy/utils/random.py (hunk 2 above)
Comment (mapazarr) [inline gammapy/utils/random.py]: It was intended for returning the same units as the input, but actually it's nonsense. After looking again to the function, I have simplified this part.

File: gammapy/utils/random.py (hunk 3 above)
Comment (mapazarr) [inline gammapy/utils/random.py]: As I said, this has been simplified. You're right no need to transform units: it was a relic of the old version. I removed it already.

File: gammapy/utils/random.py (hunk 1 above)
Comment (mapazarr) [inline gammapy/utils/random.py]: OK. Done.

**Comment by mapazarr** [comment]
I added a few more tests.

File: gammapy/utils/tests/test_random.py [hunk 4]
```diff
@@ -2,14 +2,38 @@
...
+    # test general case
     np.random.seed(0)
     lon, lat = sample_sphere(size=2)
-    assert_allclose(lon, [3.44829694, 4.49366732])
-    assert_allclose(lat, [0.20700192, 0.08988736])
+    assert_quantity_allclose(lon, Angle([3.44829694, 4.49366732], 'radian'))
+    assert_quantity_allclose(lat, Angle([0.20700192, 0.08988736], 'radian'))
+
+    # test specify a limited range
+    lon_min = Angle(40., 'degree')
```
Comment (cdeil) [inline gammapy/utils/tests/test_random.py]: Just to save a few lines, please use
```python
lon_range = Angle([40, 45], 'deg')
```
here and then pass this to `sample_sphere` directly and use  `lon_range[0]` and `lon_range[1]` in the asserts below.

File: gammapy/utils/tests/test_random.py [hunk 5]
```diff
@@ -2,14 +2,38 @@
...
+    lon_max = Angle(45., 'degree')
+    lat_min = Angle(10., 'degree')
+    lat_max = Angle(15., 'degree')
+    lon, lat = sample_sphere(size=10,
+                             lon_range=Angle([lon_min, lon_max]),
+                             lat_range=Angle([lat_min, lat_max]))
+    assert (lon_min <= lon).all() & (lon < lon_max).all()
```
Comment (cdeil) [inline gammapy/utils/tests/test_random.py]: I find this a bit more readable (avoids the duplicated `.all()`:
```python
assert np.alltrue(lon_min <= lon) & (lon < lon_max))
```
I'm not sure if either of these spellings is the best way to write this assert (i.e. short and easy to read, but still give a good message when it fails).

There's also [numpy.testing.assert_array_less](http://docs.scipy.org/doc/numpy/reference/generated/numpy.testing.assert_array_less.html), but I haven't used it.

So please take 10 minutes to thing / play with this some more, to make sure we have a good pattern for such asserts (we'll probably have 100s once we start adding lots of tests for Gammapy).
But in the end, yes, this is not very important and whatever you put, I'll merge.

File: gammapy/utils/random.py [hunk 6]
```diff
@@ -78,16 +70,16 @@ def sample_sphere(size, lon_range=None, lat_range=None, unit='rad'):
...
     lat = np.arcsin(z)
 
     # Return result
-    if unit == 'rad':
-        return lon, lat
-    elif unit == 'deg':
-        return np.degrees(lon), np.degrees(lat)
+    if lon_format == '-180_180_deg':
+        # convert back to (-180, 180) deg
+        lon -= Angle(180., 'degree')
```
Comment (cdeil) [inline gammapy/utils/random.py]: You're re-implementing longitude angle wrapping here, although it's available in Astropy:
http://astropy.readthedocs.org/en/latest/coordinates/angles.html#wrapping-and-bounds

I know you said you tried using `Longitude` objects and there was some issue.
But I'd prefer you work out that issue even if it takes an extra hour or two and we start using Longitude objects in Gammapy.
There's many places already where we need this (catalogs, observation lists) and using `Longitude` objects should make for easier / more readable code.

If you can't get this to work Ok with `Longitude`, can you post a short explanation why?
Could it be solved with a feature request for `Angle` or `Longitude` in Astropy?

In the end, if you conclude that it's just in this specific function that `Longitude` doesn't work for some reason,
OK, I'll still merge as-is.

File: CHANGES.rst [hunk 7]
```diff
@@ -54,6 +54,7 @@ Pull requests
 - Remove gammapy.shower package [#291] (Christoph Deil)
 - Add cube background model class [#299] (Manuel Paz Arribas)
 - Use assert_quantity_allclose from Astropy [#306] (Manuel Paz Arribas)
+- sample_sphere should use Quantity objects [#283] (Manuel Paz Arribas)
```
Comment (cdeil) [inline CHANGES.rst]: Change PR title and this line in the changelog to:
```
Change sample_sphere to use Angle objects
```

File: gammapy/utils/tests/test_random.py (hunk 4 above)
Comment (mapazarr) [inline gammapy/utils/tests/test_random.py]: Done.

File: gammapy/utils/random.py (hunk 6 above)
Comment (mapazarr) [inline gammapy/utils/random.py]: Ok, I gave it some more thought, and revised the docs about `Longitude` and `Latitude`. I problem I see is that the `Longitude` class, because of the wrapping method, is not very convenient for setting longitude ranges. For instance, when setting the range `(0, 360) deg`, it will be wrapped to `(0, 0) deg`!!! One would have to define the `2 \pi` interval as `(0, 360-epsilon) deg`, with `epsilon = 1e-8` for instance. And to remember this every time one defines the longitude interval is kind of hard (it's easy to forget).

In my opinion, `Longitude` does not make things easier (at least for the case of this function). With `Angle` you need to wrap angles, but you don't need to think about the wraping of the intrerval length. With `Longitude`, you don't need to wrap angles, but you need to remember to check if the interval is wrapped correctly.

@deil: please let me know your opinion on that.

File: gammapy/utils/tests/test_random.py (hunk 5 above)
Comment (mapazarr) [inline gammapy/utils/tests/test_random.py]: This is shorter:
```python
assert ((lon_range[0] <= lon) & (lon < lon_range[1])).all()
```

Comment (mapazarr) [inline gammapy/utils/tests/test_random.py]: Drawback: it doesn't show the values in case of failure. For this reason I also tested `numpy.testing.assert_array_less`, but this has the same problem as #306, an there is no good method in Astropy. I also don't think we should start reimplementing all assertions for quantities, So I'll use the above-mentioned syntax.

**Comment by mapazarr** [comment]
Ok, the new version uses `Longitude` and `Latitude`, but please look at my comments and the `TODO` list I just added at the beginning of the issue.

File: gammapy/utils/tests/test_random.py [hunk 8]
```diff
@@ -2,14 +2,55 @@
...
+    # test lon within (-180, 180) deg range
+    lon_range = Longitude([-40., 0.], 'degree', wrap_angle=Angle(180., 'degree'))
+    lon, lat = sample_sphere(size=10, lon_range=lon_range)
+    assert ((lon_range[0] <= lon) & (lon < lon_range[1])).all()
+    lat_range = Latitude([-90., 90.], 'degree')
+    assert ((lat_range[0] <= lat) & (lat < lat_range[1])).all()
+
+    # test lon range explicitly (0, 360) deg
+    epsilon = 1.e-8
+    lon_range = Longitude([0., 360.-epsilon], 'degree')
```
Comment (mapazarr) [inline gammapy/utils/tests/test_random.py]: @cdeil This way of defining the 360 deg boundary is necessary to avoid the `Longitude` wrapper to convert it to `0 deg`. I'm not sure this is better than just using `Angle` objects, even if the angles might need to be wrapped. BTW, I'm not even sure if we have to wrap them at all. Won't the method just work fine, provided the angle ranges are not broader than 360 deg? (I haven't tested it).

File: CHANGES.rst (hunk 7 above)
Comment (mapazarr) [inline CHANGES.rst]: Done (forgot to comment on this).

**Comment by mapazarr** [comment]
I am applying the fixes for issue #238 here. I started adding random_state parameters to some of the functions, and added the entry in the high-level doc.

File: gammapy/catalog/tests/test_xmatch.py [hunk 9]
```diff
@@ -10,13 +10,19 @@
...
 def test_catalog_xmatch_circle():
+    # initialise random number generator
     np.random.seed(0)
+    #rng = check_random_state(0)
+
     catalog = make_catalog_random_positions_sphere(size=100, center='Milky Way')
     catalog['Source_Name'] = ['source_{:04d}'.format(_) for _ in range(len(catalog))]
     catalog['Association_Radius'] = Angle(10 * np.random.random(len(catalog)), unit='deg')
+    #catalog['Association_Radius'] = Angle(rng.uniform(0, 10, len(catalog)), unit='deg')
+    # TODO: not working with rng!!! (but why did it work before??!!!)
```
Comment (mapazarr) [inline gammapy/catalog/tests/test_xmatch.py]: If I use `check_random_state` at this point, the result is not reproducible and the test at the end fails.

But the more strange thing is: why was it reproducible before? The test uses `random` functions: `make_catalog_random_positions_sphere`, which makes use of `sample_sphere` and `sample_sphere_distance` without passing the `seed` as a parameter.
Was the `seed` passed implicitly, when setting it at the beginning of the test?

Ideas? I guess I need to pass the `random_state` parameter to `sample_sphere` and `sample_sphere_distance` all the way through `make_catalog_random_positions_sphere`, right?

File: gammapy/astro/population/simulate.py [hunk 10]
```diff
@@ -126,21 +138,24 @@ def make_catalog_random_positions_sphere(size, center='Earth', distance=Quantity
     return table
 
 
-'''
 def make_cat_gauss_random(n_sources=100, glon_sigma=30, glat_sigma=1,
```
Comment (mapazarr) [inline gammapy/astro/population/simulate.py]: This function was placed between 2 sets of triple single quotes. Why? Was it an bug? Was the function commented out for some reason? Or are these quotes this some kind of decorator?

File: gammapy/catalog/tests/test_xmatch.py (hunk 9 above)
Comment (mapazarr) [inline gammapy/catalog/tests/test_xmatch.py]: Alternativelly: I can adapt only the functions that use a random generator, and leave the tests unchanged. This would be faster.

Comment (cdeil) [inline gammapy/catalog/tests/test_xmatch.py]: This was deterministic before because of the `np.random.seed(0)` call at the top of the `test_catalog_xmatch_circle` function. Because we were using the global numpy rng everywhere, it was deterministic.

As you say, now that we're using the pattern that all functions using rng calls take the random_state parameter, yes, you have to pass it around.

File: gammapy/astro/population/simulate.py (hunk 10 above)
Comment (cdeil) [inline gammapy/astro/population/simulate.py]: It's commented out.
Don't update it here, we might remove it.

If you want, you can add a TODO or make an issue that these functions need to be removed or improved.
(no time to deal with it at the moment)

File: CHANGES.rst [hunk 11]
```diff
@@ -54,6 +54,7 @@ Pull requests
 - Remove gammapy.shower package [#291] (Christoph Deil)
 - Add cube background model class [#299] (Manuel Paz Arribas)
 - Use assert_quantity_allclose from Astropy [#306] (Manuel Paz Arribas)
+- Change sample_sphere to use Angle objects [#283] (Manuel Paz Arribas)
```
Comment (cdeil) [inline CHANGES.rst]: Please change the title of this PR and the changelog entry to reflect the content.
Maybe "Consistent random number handling and improve sample_sphere"?

Comment (mapazarr) [inline CHANGES.rst]: Done.

File: gammapy/catalog/tests/test_xmatch.py (hunk 9 above)
Comment (mapazarr) [inline gammapy/catalog/tests/test_xmatch.py]: Done.

File: gammapy/astro/population/simulate.py (hunk 10 above)
Comment (mapazarr) [inline gammapy/astro/population/simulate.py]: Done.

**Comment by mapazarr** [comment]
@cdeil I am done with this PR. Ready for final review and evtly. merge.

**Comment by cdeil** [comment]
Thanks!

**Comment by cdeil** [comment]
I'm continuing with this a bit in #310.
